### ISSUE TITLE
repro: executes stage instead of using cached output

### ISSUE THREAD

**Issue by mstrupp** [iterative/dvc#9393]
# Bug Report

## Description

I have a pipeline with the stages `prepare`, `train` and `evaluate` and a set of hyperparameters. One of the `train` parameters is the learning rate. `train` trains a ML model and saves it to the disk.

I ran many experiments with `dvc exp run -S learning_rate=<learning rate>` using different learning rates. Then I changed my `evaluate` script (added new metrics) and wanted to re-evaluate my previously trained models. I saved the already used learning rates to a csv file and wrote a script that calls `dvc exp run -S learning_rate=<learning rate>` with these learning rates. I was expecting DVC to realize that the learning rate was already used to train a model and therefore check out the outputs of that experiment from the cache. Instead the `train` stage is executed again.

The next thing I tried to narrow down the issue was to run `dvc exp apply <exp>` for old experiments and then run `dvc status`.
I assumed that running `dvc status` right after `dvc exp apply` should never show a dependency change in the parameters.
But when I run `dvc status`, the output shows that `train.learning_rate` changed and I guess that's why the pipeline executes `train` again when I run `dvc repro`. The strange thing is that `dvc params diff <exp>` does not show any changes in the parameters (it returns an empty line). See the output below for the commands I have used.

```console
$ dvc exp apply bluer-aces
Changes for experiment 'bluer-aces' have been applied to your current workspace.

$ dvc status
train:
        changed deps:
                params.yaml:
                        modified:           train.learning_rate

$ dvc params diff bluer-aces

```

After some more investigations, I found out that DVC only re-executes `train` for certain learning rates.
These are examples for learning rates that cause a re-execution of `train`:
```console
5.73605581412296e-7
0.00004257845658686494
```
With these learning rates on the other hand, the pipeline does not re-execute `train` but checks out the outputs as expected:
```console
0.0000016145255421893676
0.0006771999543764353
1.8859564511612635e-7
0.0000012134713496975088
0.0007246348070108615
1.193189121452878e-7
0.000017731843090867312
0.0000016773367511502063
```
I can't really find a pattern for when it works and when it doesn't.

Unfortunaly, I can't reproduce the issue in a minimal example. When I create a new dvc pipeline and use the "faulty" learning rates, everything works fine and as expected.

I have many more parameters that changed but they are all of type `int` or `bool`. I first suspected a precision bug in the float 
comparism of the learning rate but I am not sure of it because I can't reproduce the error in a minimal example. It also seems really strange that `dvc status` shows a change in the learning rate but `dvc params diff` does not.

### Environment information

**Output of `dvc doctor`:**

```console
$ dvc doctor
DVC version: 2.55.0 (pip)
-------------------------
Platform: Python 3.9.16 on Windows-10-10.0.19045-SP0
Subprojects:
        dvc_data = 0.47.2
        dvc_objects = 0.21.2
        dvc_render = 0.3.1
        dvc_task = 0.2.1
        scmrepo = 1.0.2
Supports:
        http (aiohttp = 3.8.4, aiohttp-retry = 2.8.3),
        https (aiohttp = 3.8.4, aiohttp-retry = 2.8.3)
Cache types: hardlink
Cache directory: NTFS on C:\
Caches: local
Remotes: local
Workspace directory: NTFS on C:\
Repo: dvc, git
Repo.site_cache_dir: C:\ProgramData\iterative\dvc\Cache\repo\d2c2183de5f1f3360ae58fee23133f8c
```

**Comment by dberenbaum** [comment]
Thanks for reporting @mstrupp! More info on [discord](https://discord.com/channels/485586884165107732/485596304961962003/1096407381103087666). It does seem like it could be a floating point precision issue. If you are able to narrow it down further, please follow up.
